File: C/bellman-ford/gemma2-9b-it.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_NODES 100
#define INF 99999

typedef struct Edge {
    int source, destination, weight;
} Edge;

void bellmanFord(int numNodes, Edge* edges, int source) {
    int dist[MAX_NODES];
    for (int i = 0; i < numNodes; i++) {
        dist[i] = INF;
    }
    dist[source] = 0;

    for (int i = 1; i <= numNodes - 1; i++) {
        for (int j = 0; j < numNodes; j++) {
            for (int k = 0; k < numNodes; k++) {
                if (edges[k].source == j && edges[k].destination == i && dist[j] != INF && dist[j] + edges[k].weight < dist[i]) {
                    dist[i] = dist[j] + edges[k].weight;
                }
            }
        }
    }

    // Check for negative cycles
    for (int j = 0; j < numNodes; j++) {
        for (int k = 0; k < numNodes; k++) {
            if (edges[k].source == j && edges[k].destination == i && dist[j] != INF && dist[j] + edges[k].weight < dist[i]) {
                return;
            }
        }
    }
}

#define EDGES 4000
#define VERTICES 1000

int main() {
    Edge* edges = (Edge*)malloc(EDGES * sizeof(Edge));

    FILE* file = fopen("../../graph_input.txt", "r");
    if (!file) {
        free(edges);
        return 1;
    }

    for (int i = 0; i < EDGES; i++) {
        if (fscanf(file, "%d %d %d", &edges[i].source, &edges[i].destination, &edges[i].weight) != 3) {
            fclose(file);
            free(edges);
            return 1;
        }
    }

    fclose(file);

    bellmanFord(VERTICES, edges, 0);

    free(edges);
    return 0;
}

File: Java/bellman-ford/Gemma29bIt.java

import java.util.Arrays;

public class Gemma29bIt {

    public static void bellmanFord(int[][] graph, int source) {
        int V = graph.length;
        int[] dist = new int[V];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[source] = 0;

        for (int i = 1; i <= V - 1; i++) {
            for (int u = 0; u < V; u++) {
                for (int v = 0; v < V; v++) {
                    if (graph[u][v] != 0 && dist[u] != Integer.MAX_VALUE && dist[u] + graph[u][v] < dist[v]) {
                        dist[v] = dist[u] + graph[u][v];
                    }
                }
            }
        }

        // Check for negative-weight cycles
        for (int u = 0; u < V; u++) {
            for (int v = 0; v < V; v++) {
                if (graph[u][v] != 0 && dist[u] != Integer.MAX_VALUE && dist[u] + graph[u][v] < dist[v]) {
                    throw new RuntimeException("Graph contains negative-weight cycles");
                }
            }
        }
    }

    public static void main(String[] args) {
        // Example usage:
        int[][] graph = {
                {0, -1, 4, 0, 0},
                {0, 0, 0, 3, 0},
                {0, 0, 0, 0, 2},
                {0, 0, 0, 0, 0},
                {0, 0, 0, 0, 0}
        };
        int source = 0;
        bellmanFord(graph, source);
    }
}
